nrf-hal-common/src/pwm.rs

```diff
@@ -0,0 +1,996 @@
+        if self.load_seq(Seq::Seq0, &*self.duty.borrow()).is_ok() {
+            self.start_seq(Seq::Seq0);
+        }
+    }
+
+    /// Sets inverted duty cycle (15 bit) for a PWM channel.
+    pub fn set_duty_off(&self, channel: Channel, duty: u16) {
+        compiler_fence(Ordering::SeqCst);
+        self.duty.borrow_mut()[usize::from(channel)] = duty.min(self.max_duty()) | 0x8000;
+        self.one_shot();
+        self.set_load_mode(LoadMode::Individual);
+        if self.load_seq(Seq::Seq0, &*self.duty.borrow()).is_ok() {
+            self.start_seq(Seq::Seq0);
+        }
+    }
+
+    /// Returns the duty cycle value for a PWM channel.        
+    #[inline(always)]
+    pub fn duty_on(&self, channel: Channel) -> u16 {
+        self.duty_on_value(usize::from(channel))
+    }
+
+    /// Returns the inverted duty cycle value for a PWM group.
+    #[inline(always)]
+    pub fn duty_off(&self, channel: Channel) -> u16 {
+        self.duty_off_value(usize::from(channel))
+    }
+
+    /// Sets number of playbacks of sequences.
+    #[inline(always)]
+    pub fn set_loop(&self, mode: Loop) {
+        self.pwm.loop_.write(|w| match mode {
+            Loop::Disabled => w.cnt().disabled(),
+            Loop::Times(n) => unsafe { w.cnt().bits(n) },
+            Loop::Inf => unsafe { w.cnt().bits(2) },
+        });
+        self.pwm.shorts.write(|w| match mode {
+            Loop::Inf => w.loopsdone_seqstart0().enabled(),
+            _ => w.loopsdone_seqstart0().disabled(),
+        });
+    }
+
+    /// Looping disabled (stop at the end of the sequence).
+    #[inline(always)]
+    pub fn one_shot(&self) -> &Self {
+        self.set_loop(Loop::Disabled);
+        self
+    }
+
+    /// Loops playback of sequences indefinately.
+    #[inline(always)]
+    pub fn loop_inf(&self) -> &Self {
+        self.set_loop(Loop::Inf);
+        self
+    }
+
+    /// Sets number of playbacks of sequences.
+    #[inline(always)]
+    pub fn repeat(&self, times: u16) -> &Self {
+        self.set_loop(Loop::Times(times));
+        self
+    }
+
+    /// Sets number of additional PWM periods between samples loaded into compare register.
+    #[inline(always)]
+    pub fn set_seq_refresh(&self, seq: Seq, periods: u32) -> &Self {
+        match seq {
+            Seq::Seq0 => self.pwm.seq0.refresh.write(|w| unsafe { w.bits(periods) }),
+            Seq::Seq1 => self.pwm.seq1.refresh.write(|w| unsafe { w.bits(periods) }),
+        }
+        self
+    }
+
+    /// Sets number of additional PWM periods after the sequence ends.
+    #[inline(always)]
+    pub fn set_seq_end_delay(&self, seq: Seq, periods: u32) -> &Self {
+        match seq {
+            Seq::Seq0 => self.pwm.seq0.enddelay.write(|w| unsafe { w.bits(periods) }),
+            Seq::Seq1 => self.pwm.seq1.enddelay.write(|w| unsafe { w.bits(periods) }),
+        }
+        self
+    }
+
+    /// Loads a sequence buffer.
+    /// NOTE: `buf` must live until the sequence is done playing.
```

Comment:
Hmm, should this be `unsafe`? I see the same issue applies to `self.duty`. It *should* be fine from what I understand though, since all DMA does here is read.

The documentation should probably state what happens when this is not done correctly. And the same goes for functions that modify `self.duty` while PWM is running, if that makes sense.

Reply:
The worst thing that could happen if the sequence buffer doesn't live long enough is that it will play a corrupted sequence, would that be considered unsafe?

Regarding the functions that modify `self.duty`, they all quit any playing sequence and the other channels will return to their previously set duty cycle, which i think would be considered the expected behaviour? I can't see anything unsafe there, although i agree the above perhaps should be mentioned in the docs.

Reply:
Yeah, that should be fine then.

Reply:
Cool! I'll add the docs :)